Add Task Board plugin

tu2-atmanand opened the pull request:
# I am submitting a new Community Plugin

## Repo URL

<!--- Paste a link to your repo here for easy access -->
Link to my plugin:

## Release Checklist
- [x] I have tested the plugin on
  - [x]  Windows
  - [ ]  macOS
  - [x]  Linux
  - [ ]  Android _(if applicable)_
  - [ ]  iOS _(if applicable)_
- [x] My GitHub release contains all required files (as individual files, not just in the source.zip / source.tar.gz)
  - [x] `main.js`
  - [x] `manifest.json`
  - [x] `styles.css` _(optional)_
- [x] GitHub release name matches the exact version number specified in my manifest.json (_**Note:** Use the exact version number, don't include a prefix `v`_)
- [x] The `id` in my `manifest.json` matches the `id` in the `community-plugins.json` file.
- [x] My README.md describes the plugin's purpose and provides clear usage instructions.
- [x] I have read the developer policies at https://docs.obsidian.md/Developer+policies, and have assessed my plugins's adherence to these policies.
- [x] I have read the tips in https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines and have self-reviewed my plugin to avoid these common pitfalls.
- [x] I have added a license in the LICENSE file.
- [x] My project respects and is compatible with the original license of any code from other plugins that I'm using.
      I have given proper attribution to these other projects in my `README.md`.

tu2-atmanand commented:
@ObsidianReviewBot the requested changes has been made as per the Plugin Guidelines.

For the *styles via JavaScript or in HTML* issue, those css has to be hardcoded atleast for now, to not get messed up with the different themes, ill add skip for them for the next changes request.

tu2-atmanand commented:
Do I need to close and open this PR for the @ObsidianReviewBot to run the recheck again?

tu2-atmanand commented:
/skip the hardcoded styles inside HTML and Javascript are for the settings input field, which has to be kept fixed for now, due to the  layout. Different themes might mess up the layout, hence needs to be fixed, atleast for now. Ill move them to css files in future releases.

All the other request changes as per the [Plugin Guidelines](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines) have been made.

tu2-atmanand commented:
@ObsidianReviewBot all the other requested changes has been made, I guess the `Changes Requested` label can be removed now. And a developer can be assigned to review this.

tu2-atmanand commented:
/scan my repo again

tu2-atmanand commented:
@joethei can you please look into this PR ?

I have made the changes, requested by the Bot and my plugin has been properly working and has been tested using BRAT. The only remaining changes are either optional or minor. Can you please check my plugin and mention the changes ill be required to do, atleast ill start doing them.

Thank you!

joethei commented:
[.workspace-tabs .workspace-leaf .view-content {](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/styles.css#L9), [.workspace-leaf-content .view-content {](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/styles.css#L13), [.modal {](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/styles.css#L17), [.modal-close-button {](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/styles.css#L815)
Please don't overwrite core styling in a plugin, add another class/data attribute so it only applies to your plugin.

[this.createLocalStorageAndScanModifiedFiles();](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/main.ts#L77-L80)
Because these two tasks are quite heavy in calculation, they should be run after the layout is ready.
Any blocking code that is inside the regular `onload` will block the UI from being shown.

[.setViewState({ type: VIEW_TYPE_TASKBOARD, active: true });](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/main.ts#L96)
Use the new `ensureSideLeaf` function.

[id: "1",](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/main.ts#L177), [id: "2",](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/main.ts#L190), [id: "3",](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/main.ts#L199)
The command id should give a indication what the command is used for.

[this.registerEvent(](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/main.ts#L265-L283)
Remove these unused event listeners, they just add unecessary load.

[this.registerDomEvent(closeButton, "mouseenter", () => {](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/main.ts#L289-L291)
Changes should be saved right after they have been made, there are many ways to quit Obsidian without triggering this.

[const moment = require("moment");](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/utils/ScanningVault.ts#L104), [const moment = require("moment");](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/components/Column.tsx#L53)
Import it from the Obsidian package instead, and using the same syntax you import everything else.

[1: "Save",](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/utils/lang/locale/en.ts#L2-L148)
In the rest of the code I had no idea what string should be rendered, it should be possible to see just from the identifier what kind of message should be shown.
And for a lot of these english strings and many more strings found in the rest of the code: [use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

[58: "Global Settings",](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/utils/lang/locale/en.ts#L59), [36: "Plugin Global Settings",](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/utils/lang/locale/en.ts#L37), [37: "Settings",](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/utils/lang/locale/en.ts#L38), [contentEl.createEl("h1", {](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/services/TaskBoardSettingConstructUI.ts#L80-L84)
[Don't add a top-level heading in the settings tab, such as "General", "Settings", or the name of your plugin.](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Only+use+headings+under+settings+if+you+have+more+than+one+section.), 


[private static createFragmentWithHTML = (html: string) =>](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/services/TaskBoardSettingConstructUI.ts#L21-L25)
use the existing `sanitizeHTMLToDom` function instead.

[const cssStyles = ](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/services/TaskBoardSettingConstructUI.ts#L98-L118)
This should be in the `styles.css` file.

[// Setting for Plugin Language](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/services/TaskBoardSettingConstructUI.ts#L196-L215)
This should not be a setting, the plugin should always follow the language configured in Obsidian, if you don't have the translation for a particular language fallback to english.

[contentEl.createEl("h4", { text: t(79) });](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/services/TaskBoardSettingConstructUI.ts#L195), [contentEl.createEl("h4", { text: t(88) });](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/services/TaskBoardSettingConstructUI.ts#L266), [contentEl.createEl("h4", { text: t(100) });](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/services/TaskBoardSettingConstructUI.ts#L407), [contentEl.createEl("h4", {](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/services/TaskBoardSettingConstructUI.ts#L448-L450), 

[const img = createEl("img", {](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/services/TaskBoardSettingConstructUI.ts#L615-L619), [const img = createEl("img", {](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/services/TaskBoardSettingConstructUI.ts#L695-L700)
This needs to be embedded in the plugin, not loaded externally.

[placeholder="Hit Enter after typing"](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/modal/AddOrEditTaskModal.tsx#L353), [`<h2>Add Column</h2>`](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/modal/AddColumnModal.tsx#L36)
There are also some strings in your code that haven't been translated yet.

[console.log('Task deletion canceled');](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/components/Column.tsx#L90), [console.log("Board Deletion Operation Cancelled.");](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/modal/BoardConfigModal.tsx#L153)
Please avoid unnecessary logging. If you need it for debugging purposes, add a check to only log during development.

tu2-atmanand commented:
Wow, I missed many things ! Embarrassed to see so many silly mistakes of mine. Will rectify them ASAP.

> [this.registerDomEvent(closeButton, "mouseenter", () => {](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/main.ts#L289-L291)
> Changes should be saved right after they have been made, there are many ways to quit Obsidian without triggering this.

But talking about specifically the above functionality. I strongly agree that Obsidian might close unexpectedly or in a lot of other ways. In these cases, it will only give a bad user experience in a way where user won't able to see the task updated on the board while opening the plugin again. But this won't gong affect the data of user from notes, there will be absolute no data loss. And I also have an option in the **file-menu**, for user to scan updated tasks from any specific file.

But if I try to write the data every time the user makes any changes to the task or also to the file in which these tasks exist, it will write to disk for all these events. These are too many Disk read/write operations and that is the reason why I decided to go with this functionality.

I will love to hear if there is a better approach for implementing this functionality. 
In the future, I have planned to optimize the algorithm to scan whether the note content has been changed or other parts of the file has been changed in this issue, but dont have a specific solution as of now : https://github.com/tu2-atmanand/Task-Board/issues/83

tu2-atmanand commented:
[.setViewState({ type: VIEW_TYPE_TASKBOARD, active: true });](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/main.ts#L96)
Use the new ensureSideLeaf function.

I checked the ensureSideLeaf API, didn't quite understand, why you want me to open the Plugin view in sideLeaf ?
The plugin is intended to open as a normal leaf, just like a Tab. If the user wants, they can move it into the sideLeaf.

tu2-atmanand commented:
> In the rest of the code I had no idea what string should be rendered, it should be possible to see just from the identifier what kind of message should be shown.

I understand that for new contributors, they might find it challenging to understand the code and what is the message which will be shown to the users. For translation, I considered the options such as [i18n](https://www.npmjs.com/package/i18n) by referring to envelope plugin earlier. But then I felt to keep everything simple and to have full control. I decided to go with the currently used method.

For having the functionality to see the message from the identifier, I don't know how can I do it in eslint. I personally, keep the `en.ts` file open aside, to refer to the string at the current position. Will appreciate it if anyone know how to have this option during development.

tu2-atmanand commented:
> [contentEl.createEl("h4", { text: t(79) });](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/services/TaskBoardSettingConstructUI.ts#L195), [contentEl.createEl("h4", { text: t(88) });](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/services/TaskBoardSettingConstructUI.ts#L266), [contentEl.createEl("h4", { text: t(100) });](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/services/TaskBoardSettingConstructUI.ts#L407), [contentEl.createEl("h4", {](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/services/TaskBoardSettingConstructUI.ts#L448-L450),

I didnt quite understood why this is not allowed. I was actually trying to make groups for similar settings, so user can navigate easily. But I have found a better approach for using tabs, inspired from the [Folder Notes](https://github.com/LostPaul/obsidian-folder-notes/blob/c1e7567a6e83723a8a8e775f415caca09664bb91/src/settings/SettingsTab.ts). I'll implement the same in my plugin as well, for better partitioning and organizing the settings.



> [const img = createEl("img", {](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/services/TaskBoardSettingConstructUI.ts#L615-L619), [const img = createEl("img", {](https://github.com/tu2-atmanand/Task-Board/blob/2b43f981756efb365dbbf64ef2020280226a226a/src/services/TaskBoardSettingConstructUI.ts#L695-L700)
> This needs to be embedded in the plugin, not loaded externally.

Well, I agree that, in case of user is offline, these images won't be possible to fetch, which will give an error. For the buymecoffee image it will be an issue since, it shows the count. But yes, Its possible for me to convert them into svg code to embed.

**Questions :** While working on these two issue, an important question has been raised into my mind. How the management of community plugins works by Obsidian team. Because for the first issue of adding headings into setting tab, i have referred multiple plugins for the same. And for embedding the sponsorship images, i referred the [Highlitr plugin](https://github.com/chetachiezikeuzor/Highlightr-Plugin/blob/61341a464456cdc868ffa1480341136025ec8389/src/settings/settingsTab.ts), where the developers has used the same method. So, this can be done once the plugin has been published on the plugin marketplace. So is there a loop-hole here, where developers can do whatever they want in their plugins once it gets approved ?

tu2-atmanand commented:
Hii @joethei !
Thank you for the extensive review.

I have made all the requested changes and fixes. And a new version has been released with these fixes : [Task Board V1.0.2](https://github.com/tu2-atmanand/Task-Board/releases/tag/1.0.2)

The changes which I haven't made, I have explained the reason for the same in the above comments. Let me know if there are any other changes ill be required to do. Eagerly waiting to publish my plugin !

Thanks !

tu2-atmanand commented:
Also, I tried using the moment from Obsidian, but later found out that, there is issue with the implementation. I have tried the [solution given my mnaoumov](https://discord.com/channels/686053708261228577/840286264964022302/1312184581243535423). But obviously it's a workaround. I hope Obsidian team will resolve this issue and will implement the correct way to use the package.

Another favor I wanted is, with the current solution from mnaoumov, I am getting VS Code errors at present which is annoying, if I could disable them, till a better solution comes up, that will be nice. Those are only VS Code errors and I don't get them during Runtime or during build, so it's a small request, you can ignore it if the better solution comes up before that.

Thank you !

tu2-atmanand commented:
Hii @joethei 

The few changes which I postponed earlier and which were required as per the plugin guidelines have been implemented now. I hope I have solved most of the issue, if still there are any changes required, let me know.

Also, above comments had explained why I haven't made the few changes you requested me in the first review.

Following are the changes : 

- Moved all embedded styles from all my modals into the styles.css file. (The little things which are still hardcoded are required as per my feature.)
- Migrated from `vault.read()` to `vault.cachedRead()`.
- Migrated from `vault.modify()` to `vault.process()`.
- Moved few of the things inside `onLayoutReady()`.
- FIX : I was using global instance of `App` earlier. Fixed to use `plugin.app`
- Organized settings into tabs.


Here are the new features and bug fixes : https://github.com/tu2-atmanand/Task-Board/releases/tag/1.1.0

Looking forward to seeing my plugin in Obsidian Community marketplace!

joethei commented:
[await plugin.app.vault.process(file, () => newContent);](https://github.com/tu2-atmanand/Task-Board/blob/d1a36150d3136f1bc8b1ae0b34f1d8884975568d/src/utils/MarkdownFileOperations.ts#L52)
Using `vault.modify` was already correct, `vault.process` is the wrong API for what you are using.
`process` is also doing a `.read`, which is not needed here.
See this for more details: <https://youtu.be/RvtV9DKr3ag?si=r3SSpIHti9c_zYfg&t=53>

tu2-atmanand commented:
Yeah! Now I am clear about the APIs, thanks for sharing the video @joethei !

I actually used it since it has the ability to update the file in background and the atomic file operations feature. But I understand that I am actually first reading the content using some other function then doing the operations and then writing back to it using another function, so that makes sense for me to use modify. 

In the future, if I felt a need that some other plugin might mess up with my plugin operations, that time I might move on to use process directly instead of reading and writing separately.

Made the necessary changes, along with the few strings translations I missed last time. Released a new version : https://github.com/tu2-atmanand/Task-Board/releases/tag/1.1.1

joethei commented:
> Now, if I try to write the data every time the user makes any changes to the task or also to the file in which these tasks exist, it will write to disk for all these events. These are too many Disk read/write operations and that is the reason why I decided to go with this method.

That's not that many disk read/writes, it's not problematic.



> > In the rest of the code I had no idea what string should be rendered, it should be possible to see just from the identifier what kind of message should be shown.
> 
> I understand that for new contributors, they might find it challenging to understand the code and what is the message which will be shown to the users. For translation, I considered the options such as [![](data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAklEQVR4AewaftIAAACiSURBVMXBsQmDUABF0evLrwISy2yQ1g3UzpHcyBXs9HeCE2SJgEhq0waxEBTeOckEK0bCTJgJM2EW2PGoa9Ki4EpLjMxdx1ZgR1oUPJuGq81dx5YwE2bCLHDSMgxs3fOcW5ZxROCkd1Wx9ep70rLkCGEmzISZMBNmwkyYCTNhJsyEmTATZsIssGOJkTM+bct3HPm3xMieZIIVI2EmzISZMPsBPLUeCZWhvyQAAAAASUVORK5CYII=)i18n](https://www.npmjs.com/package/i18n) by referring to envelope plugin earlier. But then I felt to keep everything simple and to have full control. I decided to go with the currently used method.
> 
> For having the functionality to see the message from the identifier, I don't know how can I do it in eslint. I personally, keep the `en.ts` file open aside, to refer to the string at the current position. Will appreciate it if anyone know how to have this option during development.

Instead of `131: "Add new task in current file",`, it should be `"addNewTask": "Add new task in current file"`, for example.
To group them you can use nesting like we do in our translation files: <https://github.com/obsidianmd/obsidian-translations/blob/master/en.json>

tu2-atmanand commented:
Hi @joethei,  

### Firstly...  

I’ve given your suggestions some thought, and I completely agree with you about writing data directly to disk. Initially, I came up with the concept of using `sessionStorage` because I felt there were significantly more read operations than write operations. To address this, I wanted to create a caching mechanism. However, upon reflection, I now realize there likely won’t be excessive disk read/write operations for the `tasks.json` file.  

To optimize read operations, I considered keeping `sessionStorage` solely for reading purposes. The idea was to write data to both the disk and `sessionStorage` simultaneously while reading exclusively from `sessionStorage`. However, after further consideration, I feel this complexity isn’t necessary right now, and as a result, I’ve removed the use of `sessionStorage` entirely from the plugin.  

Regarding your second suggestion, thank you for sharing the resource! I think adopting string IDs is a great idea, and I have updated my project accordingly.  

---

### Secondly...  

I’d like to apologize for making a few additional changes beyond the ones mentioned above. These changes were necessary because I identified a significant flaw in my plugin’s UI performance. Previously, I was passing the entire data set as a prop to the `Column` component, where it was processed to determine which tasks belonged to each column. This approach caused performance bottlenecks.  

To address this, I’ve implemented major performance improvements that have drastically enhanced the UI’s responsiveness and reduced resource utilization.  

I kindly request you to review the new changes. I wanted to help some way, so provided the list of changes I have made :  

- **React Memoization**: Implemented memoization across all three components (`KanbanBoard.tsx`, `Column.tsx`, and `TaskItem.tsx`).  
- **Removed React Dispatcher**: Simplified state management by removing React Dispatcher.  
- **Updated AddColumnModal**: Converted `AddColumnModal` from a functional component (`React.FC`) to a class component.  
- **Optimized Data Processing**: Data is now pre-processed in `KanbanBoard.tsx` before being passed to each column component (`Column.tsx`) for rendering.  
- **TaskItem State Optimization**: Removed state variables inside the `TaskItem` component, which resolved the issue : #53  
- **On-Demand Rendering**: The description section for each `TaskItem` card is now only rendered when the user expands it, significantly reducing the use of Obsidian's render API.  
- **Animations**: Added animations to `TaskItem` components using the **framer-motion** package.  
- **Fixed Component Usage**: Replaced the incorrect usage of `new Component` with `ItemView` as a component.  

- **Removed sessionStorage**: Completely removed `sessionStorage` and switched to direct disk read/write operations.  
- **Improved Language Translation**: Updated language translation approach to use string keys over numeric keys.

---

### Lastly...  

Thank you for your understanding and patience with these changes. I genuinely appreciate your time and effort in reviewing my plugin. I’m excited to see it make its way into the Community Plugins marketplace soon!  

A new version has been released with all these updates : https://github.com/tu2-atmanand/Task-Board/releases/tag/1.2.0

Additionally, after thorough testing, I’ve created an introductory video for the plugin : [Task Board plugin - Introduction](https://youtu.be/ZizsPBuXW8g).
